main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1077 @@
+          fparamKenv => acc ++ fparamKenv
+        }
+      }
+      val schemeKenvVal = inferScheme(sc, root)
+      val effKenvVal = inferType(eff, KindMatch.subKindOf(Kind.Bool), root)
+
+      flatMapN(fparamKenvVal, schemeKenvVal, effKenvVal) {
+        case (fparamKenv, schemeKenv, effKenv) => KindEnv.merge(fparamKenv, schemeKenv, effKenv)
+      }
+
+  }
+
+  /**
+    * Infers a kind environment from the given formal param.
+    */
+  private def inferFparam(fparam0: ResolvedAst.FormalParam, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = fparam0 match {
+    case ResolvedAst.FormalParam(_, _, tpe0, _) => inferType(tpe0, KindMatch.subKindOf(Kind.Star), root)
+  }
+
+  /**
+    * Infers a kind environment from the given scheme.
+    */
+  private def inferScheme(sc0: ResolvedAst.Scheme, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = sc0 match {
+    case ResolvedAst.Scheme(_, constraints, base) =>
+      val baseKenvVal = inferType(base, KindMatch.subKindOf(Kind.Star), root)
+      val tconstrsKenvsVal = Validation.traverse(constraints)(inferTconstr(_, root))
+
+      Validation.flatMapN(baseKenvVal, tconstrsKenvsVal) {
+        case (baseKenv, tconstrKenvs) => Validation.fold(tconstrKenvs, baseKenv)(_ ++ _)
+      }
+  }
+
+  /**
+    * Infers a kind environment from the given type constraint.
+    */
+  private def inferTconstr(tconstr: ResolvedAst.TypeConstraint, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = tconstr match {
+    case ResolvedAst.TypeConstraint(clazz, tpe, loc) =>
+      val kind = getClassKind(root.classes(clazz))
+      inferType(tpe, KindMatch.subKindOf(kind), root)
+  }
+
+  /**
+    * Infers a kind environment from the given type, with an expectation from context.
+    */
+  private def inferType(tpe: UnkindedType, expectedType: KindMatch, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = tpe.baseType match {
+    // Case 1: the type constructor is a variable: all args are * and the constructor is * -> * -> * ... -> expectedType
+    case tvar: UnkindedType.Var =>
+      val kind = tpe.typeArguments.foldLeft(expectedType.kind) {
+        case (acc, _) => Kind.Star ->: acc
+      }
+
+      Validation.fold(tpe.typeArguments, KindEnv.singleton(tvar -> kind)) {
+        case (acc, targ) => flatMapN(inferType(targ, KindMatch.subKindOf(Kind.Star), root)) {
+          kenv => acc ++ kenv
+        }
+      }
+    case UnkindedType.Cst(cst, loc) =>
+      val tyconKind = getTyconKind(cst, root)
+      val args = Kind.args(tyconKind)
+
+      Validation.fold(tpe.typeArguments.zip(args), KindEnv.empty) {
+        case (acc, (targ, kind)) => flatMapN(inferType(targ, KindMatch.subKindOf(kind), root)) {
+          kenv => acc ++ kenv
+        }
+      }
+    case UnkindedType.Apply(_, _) => throw InternalCompilerException("Unexpected type application.")
```

Comment:
Oooh okay. Why is this safe here? Also Exceptions are usually handled at the bottom.

Reply:
This is safe because of the definition of `UnkindedType.baseType`: on an `Apply` it just recurses. I guess I could enforce this in the type system with a union trait.